Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py

```diff
@@ -0,0 +1,714 @@
+            # might not be true for all components due to presence of monitors etc.
+            self.get_detid_array = self._get_detid_array_rect_detector
+        else:
+            self.get_detid_array = self._get_detid_array_comp_assembly
+
+    @staticmethod
+    def _split_string_trailing_int(name):
+        *_, match = re.finditer(r'\d+', name)  # last match
+        return name[:match.start()], name[match.start():]  # prefix, num_str
+
+    @staticmethod
+    def _find_nearest_child_to_component(parent, comp, excl=None, ndepth=1):
+        dist = np.inf
+        nearest_child = None
+        for ichild in range(parent.nelements()):
+            child = parent[ichild]
+            if hasattr(child, 'nelements'):
+                while child.nelements() == 1:
+                    child = child[0]
+            is_correct_type = 'CompAssembly' in child.type() or child.type() == 'DetectorComponent'
+            is_not_excluded = excl is None or child.getFullName() != excl
+            if is_correct_type and is_not_excluded:
+                this_dist = child.getDistance(comp)
+                if this_dist < dist:
+                    dist = this_dist
+                    nearest_child = child
+        if ndepth > 1:
+            nearest_child = InstrumentArrayConverter._find_nearest_child_to_component(nearest_child, comp,
+                                                                                      ndepth=ndepth - 1)
+        return nearest_child
+
+    def _find_nearest_adjacent_col_component(self, dcol, col, ncols, bank_name, col_prefix, col_str, delim, ncols_edge):
+        next_col_prefix, next_col_str = None, None
+        isLHS = np.any(dcol + col <= ncols_edge)
+        isRHS = np.any(dcol + col >= ncols - ncols_edge + 1)
+        if isLHS or isRHS:
+            # get tube/col at each end of detector to get tube separation
+            detLHS = self.inst.getComponentByName(
+                delim.join([bank_name, f'{col_prefix}{1:0{len(col_str)}d}']))
+            detRHS = self.inst.getComponentByName(
+                delim.join([bank_name, f'{col_prefix}{ncols:0{len(col_str)}d}']))
+            det_sep = detLHS.getDistance(detRHS) / (ncols - 1)  # loose half pix at each end (cen-cen dist.)
+            # look for adjacent tube in adjacent banks
+            det_ref = detLHS if isLHS else detRHS
+            next_det = self._find_nearest_child_to_component(self.inst, det_ref, excl=bank_name, ndepth=2)
+            if next_det is not None and next_det.getDistance(det_ref) < 1.1 * det_sep:
+                # is considered adjacent
+                next_col_prefix, next_col_str = self._split_string_trailing_int(next_det.getFullName())
+        return next_col_prefix, next_col_str, isLHS, isRHS
+
+    def _get_detid_array_comp_assembly(self, bank, detid, row, col, drows, dcols, nrows_edge, ncols_edge=1):
+        ndet_per_spec = len(self.ws.getSpectrum(self.ws.getIndicesFromDetectorIDs([detid])[0]).getDetectorIDs())
```

Comment:
Please can you separate this line out a bit. If I run the test script on the PR with InputWorkspace set to a workspace other than SXD23767 (eg one from another instrument) I get an "index out of range" error from this line. This happens because self.ws.getIndicesFromDetectorIDs([detid]) returns an empty list. In that case I think this function should log an error and then the for loop in exec should carry on with the next peak

Reply:
So passing the error back up the function chain is a bit messy, so instead I have added additional validation and checks at a higher level: in the input validation I check the peak and data workspaces have the same instrument and in the exec loop over all peaks I check the detector of each peak has a valid component index (i.e. exists in the instrument) and that the detector is not a monitor or masked. I think this will fix the error you found.